cardano-testnet/src/Testnet/Components/DReps.hs

```diff
@@ -100,50 +119,148 @@ generateRegistrationCertificate execConfig work prefix drepKeyPair depositAmount
+--
+-- Returns a list of generated @File VoteFile In@ representing the paths to
+-- the generated voting files.
+generateVoteFiles :: (MonadTest m, MonadIO m, MonadCatch m)
+  => H.ExecConfig
+  -> FilePath
+  -> String
+  -> String
+  -> Word32
+  -> [(PaymentKeyPair, [Char])]
+  -> m [File VoteFile In]
+generateVoteFiles execConfig work prefix governanceActionTxId governanceActionIndex allVotes = do
+  baseDir <- H.createDirectoryIfMissing $ work </> prefix
+  forM (zip [(1 :: Integer)..] allVotes) $ \(idx, (drepKeyPair, vote)) -> do
+    let path = File (baseDir </> "vote-" <> show idx)
+    void $ H.execCli' execConfig
+      [ "conway", "governance", "vote", "create"
+      , "--" ++ vote
+      , "--governance-action-tx-id", governanceActionTxId
+      , "--governance-action-index", show @Word32 governanceActionIndex
+      , "--drep-verification-key-file", paymentVKey drepKeyPair
+      , "--out-file", unFile path
+      ]
+    return path
+
+-- | Composes a decentralized representative (DRep) voting transaction body
+-- (without signing) using @cardano-cli@.
+--
+-- This function takes seven parameters:
+--
+-- * 'execConfig': Specifies the CLI execution configuration.
+-- * 'epochStateView': Current epoch state view for transaction building. It can be obtained
+--                     using the 'getEpochStateView' function.
+-- * 'sbe': The Shelley-based era (e.g., 'ShelleyBasedEraShelley') in which the transaction will be constructed.
+-- * 'work': Base directory path where the transaction body file will be stored.
+-- * 'prefix': Prefix for the output transaction body file name. The extension will be @.txbody@.
+-- * 'votes': List of voting files (@File VoteFile In@) to include in the transaction,
+--            obtained using 'generateVoteFiles'.
+-- * 'wallet': Payment key information associated with the transaction,
+--             as returned by 'cardanoTestnetDefault'.
+--
+-- Returns the generated @File TxBody In@ file path to the transaction body.
+createVotingTxBody
+  :: (H.MonadAssertion m, MonadTest m, MonadCatch m, MonadIO m)
+  => H.ExecConfig
+  -> EpochStateView
+  -> ShelleyBasedEra era
+  -> FilePath
+  -> String
+  -> [File VoteFile In]
+  -> PaymentKeyInfo
+  -> m (File TxBody In)
+createVotingTxBody execConfig epochStateView sbe work prefix votes wallet = do
+  let dRepVotingTxBody = File (work </> prefix <> ".txbody")
+  walletLargestUTXO <- findLargestUtxoForPaymentKey epochStateView sbe wallet
+  void $ H.execCli' execConfig $
+    [ "conway", "transaction", "build"
+    , "--change-address", Text.unpack $ paymentKeyInfoAddr wallet
+    , "--tx-in", Text.unpack $ renderTxIn walletLargestUTXO
+    ] ++ (concat [["--vote-file", voteFile] | File voteFile <- votes]) ++
+    [ "--witness-override", show @Int (length votes)
+    , "--out-file", unFile dRepVotingTxBody
+    ]
+  return dRepVotingTxBody
+
 -- Transaction signing
 
 data SignedTx
 
+class KeyPair a where
```

Comment:
Nice idea. I think this should be moved to `Testnet.Runtime` to make it more accessible.

In general I think we could create a `KeyPair` type with a phantom variable which would represent keys and just use uninhabited types to mark the types. I mean, instead of:
```haskell
data PaymentKeyPair = PaymentKeyPair
  { paymentVKey :: FilePath
  , paymentSKey :: FilePath
  } deriving (Eq, Show)

data StakingKeyPair = StakingKeyPair
  { stakingVKey :: FilePath
  , stakingSKey :: FilePath
  } deriving (Eq, Show)
```
we could have
```haskell
data Staking
data Payment

data KeyPair (type :: Type) = KeyPair
  { vKey :: File type InOut
  , sKey :: File type InOut
  }
```
and just use `KeyPair Staking` and `KeyPair Payment`. But that's out of scope of this PR.

Reply:
100%